awx/main/access.py

```diff
@@ -772,13 +771,23 @@ def can_delete(self, obj):
         return True
 
     def can_attach(self, obj, sub_obj, relationship, *args, **kwargs):
+        # If the request is updating the membership, check the membership role permissions instead
+        if relationship in ('member_role.members', 'admin_role.members'):
+            rel_role = getattr(obj, relationship.split('.')[0])
+            return RoleAccess(self.user).can_attach(rel_role, sub_obj, 'members', *args, **kwargs)
```

Comment:
Attachment to roles is pretty fundamental to the app, and I don't want to assume there are not other cases of this. So I think this is a very good candidate to put in `BaseAccess` instead. What would you think about

```diff
diff --git a/awx/main/access.py b/awx/main/access.py
index 250ce99a66..a34e46e113 100644
--- a/awx/main/access.py
+++ b/awx/main/access.py
@@ -235,6 +235,10 @@ class BaseAccess(object):
 
     def can_attach(self, obj, sub_obj, relationship, data,
                    skip_sub_obj_read_check=False):
+        if relationship.split('.')[-1] == 'members' and relationship.count('.') > 0:
+            rel_role = getattr(obj, relationship.split('.')[0])
+            return RoleAccess(self.user).can_attach(rel_role, sub_obj, 'members', data,
+                                                    skip_sub_obj_read_check=skip_sub_obj_read_check)
         if skip_sub_obj_read_check:
             return self.can_change(obj, None)
         else:
```

Reply:
Hm, quite possibly.

Reply:
On second thought, I'm not comfortable making such an extensive change without a better understanding of the implications.

Reply:
That's fine, but I would still much prefer this long term. We could protect against false positives by inspecting the field `obj._meta.get_field(relationship.split('.')[0])`

Reply:
Definitely.  If such relationships are recorded in the database via role membership, then ultimately it's the right thing to do.

Reply:
Thinking more on this, `grep -r "relationship =" awx/api` tells me that the 3 cases you are handling here are the only actual cases of an "extra hop", in addition to the parents case for teams already dealt with.